motoman_driver/src/industrial_robot_client/joint_trajectory_streamer.cpp

```diff
@@ -159,6 +159,73 @@ void JointTrajectoryStreamer::jointTrajectoryCB(const trajectory_msgs::JointTraj
   send_to_robot(new_traj_msgs);
 }
 
+void JointTrajectoryStreamer::jointCommandCB(const trajectory_msgs::JointTrajectoryConstPtr &msg)
+{
+  // read current state value (should be atomic)
+  int state = this->state_;
+
+  ROS_DEBUG("Current state is: %d", state);
+
+  //If current state is idle, set to POINT_STREAMING
+  if (TransferStates::IDLE == state)
+  {
+    this->mutex_.lock();
+    this->state_ = TransferStates::POINT_STREAMING;
+    this->current_point_ = 0;
+    this->streaming_sequence = 0;
+    this->streaming_start_ = ros::Time::now();
+    this->mutex_.unlock();
+    state = TransferStates::POINT_STREAMING;
+    ROS_INFO("First joint point received. Starting on-the-fly streaming.");
+  }
+
+  //if current state is POINT_STREAMING, process incoming point.
+  if (TransferStates::POINT_STREAMING == state)
+  {
+    if (msg->points.empty())
+    {
+      ROS_INFO("Empty point received, cancelling current trajectory");
+      return;
+    }
+
+    //Else, Push point into queue
+    SimpleMessage message;
+    trajectory_msgs::JointTrajectoryPoint rbt_pt, xform_pt;
+
+    // select / reorder joints for sending to robot
+    if (!select(msg->joint_names, msg->points[0], this->all_joint_names_, &rbt_pt))
```

Comment:
Does `select(..)` also check that all required joints are present in the incoming `msg`?

If not, such a check would probably need to be added.

Reply:
The function ``select(.)`` internally checks whether any element of ``msg->joint_names`` is empty or whether a matching "ROS element" can be found for the name of each joint in the message. If either of those two conditions are false, the function returns ``false``. So I think that would be the check for all required joints being present.

Reply:
> The function `select(.)` internally checks whether any element of `msg->joint_names` is empty or whether a matching "ROS element" can be found for the name of each joint in the message. If either of those two conditions are false, the function returns `false`. So I think that would be the check for all required joints being present.

I agree with you on this. I think checking the return on select is adequate.